### PR TITLE
Somalier relate

### PR DESCRIPTION
**Pull request by ashotmarg**
## PR checklist

Closes #XXX 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

**Comment by jfy133** [comment]
Looks rather incomplete due to boilerplate containers/conda env URLS, and also has random (output?) files in the test directory - so I'm converting to draft for now!

**Comment by ashotmarg** [comment]
Hi James, thanks for going through the code.

I have actually left the default containers/conda env URLS in the code "on purpose" to emphasise that there are no biocontainer or conda images for this (somalier) program. Only docker is available.
Re the random (output?) files: these are actually the test input files (both required and optional files). These are not available in the "test_data.config" so I put them there.

So please advise how to proceed for these two points and I will fix them accordingly.

**Comment by ashotmarg** [comment]
> Is there no chance they Somalier can go on conda/biocontainers?
I am not really sure, I don't have experience with this. Can we use the module with only docker? I guess it's not ideal but maybe possible, could you check with the other modules? Alternatively, do you know what's the easiest way to push/publish a e.g., singularity image to depot.galaxyproject.org/singularity ?
I could technically make a dockerfile for biocontainers, and push the docker image of Somalier to "quay.io" (which as far as I've read is preferred over the docker hub). Then, depot.galaxyproject.org/singularity should be able to automatically (?) get/make the singularity images from the biocontainers, mirroring the biocontainers, is that correct?

I will add the test files to the test-datasets repo where relevant. E.g., the input files for this module are the output files from https://github.com/nf-core/modules/pull/1986 , so cannot really use the existing files.

**Comment by jfy133** [comment]
Ok I took a stab last night and putting on bioconda was actually rather painless!

https://github.com/bioconda/bioconda-recipes/pull/36877

I also ran the 'functional tests' in somalier's scripts directory and it appeared to work as well with no failures

Once reviewed and merged it'll be a few days and then the biocontainers  will be available :)

**Comment by jfy133** [comment]
> I will add the test files to the test-datasets repo where relevant. E.g., the input files for this module are the output files from https://github.com/nf-core/modules/pull/1986 , so cannot really use the existing files.

Also, OK - then we can focus on getting extract in first :), can move discussion over there

**Comment by ashotmarg** [comment]
Hi @jfy133, please double check this one as well, I've changed it according to the other module, so this should be (almost) ready. Re the input files that are not in the official datasets, should we keep them in the "delete_me" folder for now?

File: modules/somalier/relate/main.nf [hunk 1]
```diff
@@ -0,0 +1,46 @@
+process SOMALIER_RELATE {
+    tag "somalier_relate_all"
```
Comment (jfy133) [inline modules/somalier/relate/main.nf:2]:
```suggestion
    tag "$meta.id"
```

File: modules/somalier/relate/main.nf [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    script:
+    def args = task.ext.args ?: ''
+    def input_list = extract.collect{"$it"}.join(' ')
+    def prefix = task.ext.prefix ?: ""
+    def sample_groups_command = sample_groups ? "-g $sample_groups" : ""
+    def ped_command = ped ? "-p $ped" : ""
+
+    """
+    somalier relate \\
+    -o ${prefix}somalier.res \\
```
Comment (jfy133) [inline modules/somalier/relate/main.nf:33]:
```suggestion
    -o ${prefix} \\
```
https://nf-co.re/docs/contributing/modules#naming-conventions point 5

File: modules/somalier/relate/main.nf [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    input:
+    tuple val(meta), path(extract)
+    path(sample_groups)
+    path(ped)
+
+    output:
+    tuple val(meta), path("*somalier.res.html"),          emit: html
+    tuple val(meta), path("*somalier.res.pairs.tsv"),     emit: pairsTSV
+    tuple val(meta), path("*somalier.res.samples.tsv"),   emit: samplesTSV
```
Comment (jfy133) [inline modules/somalier/relate/main.nf:18]:
```suggestion
    tuple val(meta), path("*.html"),          emit: html
    tuple val(meta), path("*.pairs.tsv"),     emit: pairs_tsv
    tuple val(meta), path("*.samples.tsv"),   emit: samples_tsv
```

File: modules/somalier/relate/main.nf [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    def sample_groups_command = sample_groups ? "-g $sample_groups" : ""
+    def ped_command = ped ? "-p $ped" : ""
+
+    """
+    somalier relate \\
+    -o ${prefix}somalier.res \\
+    ${input_list} \\
+    ${args} \\
+    ${sample_groups_command} \\
+    ${ped_command}
```
Comment (jfy133) [inline modules/somalier/relate/main.nf:37]:
```suggestion
        -o ${prefix}somalier.res \\
        ${input_list} \\
        ${args} \\
        ${sample_groups_command} \\
        ${ped_command}
```

File: modules/somalier/relate/main.nf [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+
```
Comment (jfy133) [inline modules/somalier/relate/main.nf:39]:
```suggestion
```

File: modules/somalier/relate/meta.yml [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+      doi: "https://doi.org/10.1186/s13073-020-00761-2"
+      licence: "MIT License"
+
+input:
+  - meta:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - somalierExtract:
```
Comment (jfy133) [inline modules/somalier/relate/meta.yml:30]:
```suggestion
  - extract:
```

File: modules/somalier/relate/meta.yml [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+      pattern: "versions.yml"
+  - html:
+      type: file
+      description: html file
+      pattern: "*.{.html}"
+  - pairs:
+      type: file
+      description: tsv file with output stats for pairs of samples
+      pattern: "*.{.pairs.tsv}"
+  - samples:
```
Comment (jfy133) [inline modules/somalier/relate/meta.yml:48]:
```suggestion
  - samples_tsv:
```

File: modules/somalier/relate/meta.yml [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+output:
+  - versions:
+      type: file
+      description: File containing software versions
+      pattern: "versions.yml"
+  - html:
+      type: file
+      description: html file
+      pattern: "*.{.html}"
+  - pairs:
```
Comment (jfy133) [inline modules/somalier/relate/meta.yml:44]:
```suggestion
  - pairs_tsv:
```

**Review by jfy133 — CHANGES_REQUESTED** [review]
You're missing docs in meta.yml for `sample_groups`, `ped`.

yes keeping in `delete_me` is fine!

**Review by jfy133 — APPROVED** [review]

**Pull request merged**